File: lunar_lander_agent_base.py
import numpy as np

# np.random.seed(0)


# The resolution of the observation space = Felbontas
# The four variables of the observation space, from left to right:
#   0: X component of the vector pointing to the middle of the platform from the lander
#   1: Y component of the vector pointing to the middle of the platform from the lander
#   2: X component of the velocity vector of the lander
#   3: Y component of the velocity vector of the lander
OBSERVATION_SPACE_RESOLUTION = [31, 11, 15, 15]  # TargetX, TargetY, VelocityX, VelocityY


class LunarLanderAgentBase:
    def __init__(self, observation_space, action_space, n_iterations):
        # self.temp_q_table = []
        self.observation_space = observation_space  # [0][0] : TargetX min , [0][1] : TargetX max , [1][0] : TargetY min
        self.q_table = np.zeros(
            [*OBSERVATION_SPACE_RESOLUTION, len(action_space)])  # [TargetX][TargetY][VelocityX][VelocityY][Action]
        self.env_action_space = action_space  # elvegezheto cselekvesek
        self.n_iterations = n_iterations

        self.epsilon = 1.  # felfedezes eselye
        self.iteration = 0
        self.test = False  # teszteles elesben, itt mar nem szabad felfedezni

        # self.alpha = 0.7  # ilyen aranyban irja be a visszajelzest a q-tablaba


    @staticmethod
    def quantize_state(observation_space, state):  # allapot egesz reszre kerekitese
        targetX = round((state[0] + 300) / 20)
        # targetY = round((state[1]) / 20)
        targetY = round(np.log2(state[1] + 1) * 10 / 7.64)
        velocityX = round((state[2] + 7))
        velocityY = round((state[3] + 7))
        return targetX, targetY, velocityX, velocityY

    def epoch_end(self, epoch_reward_sum):  # minden menet utan meghivott tanulo fv
        pass
        # self.epsilon = self.epsilon / 1.001
        # for element in self.temp_q_table:
        #     element[2] += epoch_reward_sum / 10
        #
        #     quantized_state = self.quantize_state(self.observation_space, element[0])
        #
        #     self.q_table[quantized_state][element[1]] = \
        #         self.q_table[quantized_state][element[1]] * (1 - self.alpha) + \
        #         element[2] * self.alpha

    def learn(self, old_state, action, new_state, reward):  # minden lepes utan meghivott tanulo fv
        quantized_state = self.quantize_state(self.observation_space, old_state)

        self.q_table[quantized_state[0]][quantized_state[1]][quantized_state[2]][quantized_state[3]][action] = reward

        # self.temp_q_table.append([old_state, action, reward])

    def train_end(self):  # teljes tanulas vege

        # self.q_table = None  # TODO
        self.test = True

File: LunarLanderAgentBase.java
import java.util.Arrays;

public class LunarLanderAgentBase {
	
	static int tXMax = 40;
	static int tYMax = 40;
	static int vXMax = 21;
	static int vYMax = 21;
	
    // The resolution of the observation space
    // The four variables of the observation space, from left to right:
    //   0: X component of the vector pointing to the middle of the platform from the lander
    //   1: Y component of the vector pointing to the middle of the platform from the lander
    //   2: X component of the velocity vector of the lander
    //   3: Y component of the velocity vector of the lander
    static final int[] OBSERVATION_SPACE_RESOLUTION = {tXMax + 1, tYMax + 1, vXMax + 1, vYMax + 1};

    final double[][] observationSpace;
    double[][][][][] qTable;
    final int[] envActionSpace;
    private final int nIterations;

    double epsilon = 1.0;
    int iteration = 0;
    boolean test = false;

    // your variables here
    // ...
    double[][][][][] bestTable;
    double bestReward = -200;
    double lastReward = -200;

    double alpha = 0.1;
    double gamma = 0.6;
    int epsilon_step = 100;
    double epsilon_decay = 0.99;
    int save_interval = 1000;
    
    int epoch = 0;

    public LunarLanderAgentBase(double[][] observationSpace, int[] actionSpace, int nIterations) {
        this.observationSpace = observationSpace;
        this.qTable =
                new double[OBSERVATION_SPACE_RESOLUTION[0]]
                        [OBSERVATION_SPACE_RESOLUTION[1]]
                        [OBSERVATION_SPACE_RESOLUTION[2]]
                        [OBSERVATION_SPACE_RESOLUTION[3]]
                        [actionSpace.length];
        this.envActionSpace = actionSpace;
        this.nIterations = nIterations;
    }

    public static int[] quantizeState(double[][] observationSpace, double[] state) {
    	int[] res = new int[4];
    	
    	//res[0] = (int) ((state[0] + 300) / 10);
    	res[0] = (int) (Math.pow(state[0], (3.0 / 5.0)) * (tXMax / 2 / 30.63887) + tXMax / 2);
    	if (state[0] > -15 && state[0] < 15)
    		res[0] = 20;
    	
        //res[1] = (int) ((state[1]) / 4);
        res[1] = (int) (Math.log(state[1] / 100 + 1) * tYMax / 1.098612);
        if (state[1] < 0)
        	res[1] = 0;
        else if (state[1] > 180)
        	res[1] = tYMax;
        
        //res[2] = (int) ((state[2] + 7) * vXMax / 14);
    	res[2] = (int) (Math.pow(state[2], (3.0 / 5.0)) * (vXMax / 2 / 3.21409585) + vXMax / 2);

        //res[3] = (int) ((state[3] + 7) * vYMax / 14);
        res[3] = (int) (Math.pow(state[3], (3.0 / 5.0)) * (vYMax / 2 / 3.21409585) + vYMax / 2);
        
        return res;
    }

    public void epochEnd(double epochRewardSum) {
    	return;
    }

    public void learn(double[] oldState, int action, double[] newState, double reward) {

        int[] quantizedOldState = quantizeState(observationSpace, oldState);
        
        double newReward = (1 - alpha) * oldState[action] + alpha * (reward + gamma * maxQ(newState));

        qTable[quantizedOldState[0]][quantizedOldState[1]][quantizedOldState[2]][quantizedOldState[3]][action] = newReward;
    }

    private double maxQ(double[] state) {
    	
    	int[] quantized = quantizeState(observationSpace, state);
    	
    	double max = qTable[quantized[0]][quantized[1]][quantized[2]][quantized[3]][0];
		for (int i = 1; i < state.length; i++) {
			if (qTable[quantized[0]][quantized[1]][quantized[2]][quantized[3]][i] > max)
				max = qTable[quantized[0]][quantized[1]][quantized[2]][quantized[3]][i];
		}
		return max;
	}

	public void trainEnd() {
        test = true;
    }
}
